File: python/soup/soup.py
from abc import ABCMeta, abstractmethod

class Device(metaclass=ABCMeta):
    @abstractmethod
    def op1(self, param):
        pass


class DeviceA(Device, metaclass=ABCMeta):
    @abstractmethod
    def op2(self, param):
        pass


class DeviceB(Device, metaclass=ABCMeta):
    @abstractmethod
    def op2(self, x, y):
        pass


class Vendor(metaclass=ABCMeta):
    @abstractmethod
    def createDeviceA(self, index):
        pass

    @abstractmethod
    def createDeviceB(self, index):
        pass

File: rust/src/soup.rs

pub trait Device {
    fn op1(&mut self, param: i32);
}


pub trait DeviceA : Device {
    fn op2(&mut self, param: i32) -> i32;
}


pub trait DeviceB : Device {
    fn op2(&mut self, x: i32, y: i32) -> i32;
}


pub trait Vendor {
    fn create_devicea(&self, index: i32) -> Box<DeviceA + 'static>;
    fn create_deviceb(&self, index: i32) -> Box<DeviceB + 'static>;
}
